fix(messenger): make sure chats have an unread count of 0 for channels you can't view

jrainville opened the pull request:
Fixes https://github.com/status-im/status-desktop/issues/14421

The problem is that you can receive messages to  a channel, then later, before marking them as read, a permission is added to them, so you no longer have access. Then, you can't even mark it as read if it's hidden. Here, I fix it by setting the unread count on Init at 0 if the user doesn't have view access to it.

It's not the cleanest solution, but it's the simplest one to fix previous accounts that had that issue.

Another solution would be to change the DB value of the chat when we receive a new permission and detect that we no longer have view access. It's cleaner, but it doesn't fix previous account having the issue.
Also, keeping the DB value means that if the channel goes back to visible to us, we'll have a badge for the previous unread messages we had.

Let me know what you guys think.

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (21)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 0e244ed692237bccff0d2a087f14fab27959283d | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/1/) | 2024-04-16 20:07:52 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/1//consoleText) |
| :heavy_check_mark: | 0e244ed692237bccff0d2a087f14fab27959283d | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5062/1/) | 2024-04-16 20:10:44 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240416-200640-0e244ed692237bccff0d2a087f14fab27959283d-pr5062.zip) |
| :heavy_check_mark: | 0e244ed692237bccff0d2a087f14fab27959283d | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5062/1/) | 2024-04-16 20:11:30 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240416-200635-0e244ed692237bccff0d2a087f14fab27959283d-pr5062.zip) |
| :heavy_check_mark: | 0e244ed692237bccff0d2a087f14fab27959283d | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5062/1/) | 2024-04-16 20:11:57 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240416-200635-0e244ed692237bccff0d2a087f14fab27959283d-pr5062.aar) |
| | | | | | | |
| :heavy_multiplication_x: | a8000ed3b155c524d002ce76dc44c56894767d5b | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/2/) | 2024-04-17 18:39:09 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/2//consoleText) |
| :heavy_check_mark: | a8000ed3b155c524d002ce76dc44c56894767d5b | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5062/2/) | 2024-04-17 18:40:38 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240417-183759-a8000ed3b155c524d002ce76dc44c56894767d5b-pr5062.zip) |
| :heavy_check_mark: | a8000ed3b155c524d002ce76dc44c56894767d5b | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5062/2/) | 2024-04-17 18:41:42 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240417-183759-a8000ed3b155c524d002ce76dc44c56894767d5b-pr5062.zip) |
| :heavy_check_mark: | a8000ed3b155c524d002ce76dc44c56894767d5b | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5062/2/) | 2024-04-17 18:43:19 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240417-183755-a8000ed3b155c524d002ce76dc44c56894767d5b-pr5062.aar) |
| | | | | | | |
| :heavy_multiplication_x: | f142ea04e243a58f6e67b1c97524a5c007d82183 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/3/) | 2024-04-19 14:15:20 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/3//consoleText) |
| :heavy_check_mark: | f142ea04e243a58f6e67b1c97524a5c007d82183 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5062/3/) | 2024-04-19 14:16:53 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240419-141212-f142ea04e243a58f6e67b1c97524a5c007d82183-pr5062.zip) |
| :heavy_check_mark: | f142ea04e243a58f6e67b1c97524a5c007d82183 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5062/3/) | 2024-04-19 14:16:54 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240419-141212-f142ea04e243a58f6e67b1c97524a5c007d82183-pr5062.zip) |
| :heavy_check_mark: | f142ea04e243a58f6e67b1c97524a5c007d82183 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5062/3/) | 2024-04-19 14:18:12 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240419-141212-f142ea04e243a58f6e67b1c97524a5c007d82183-pr5062.aar) |
| | | | | | | |
| :heavy_check_mark: | 6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5062/4/) | 2024-04-19 16:14:01 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240419-161123-6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae-pr5062.aar) |
| :heavy_check_mark: | 6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5062/4/) | 2024-04-19 16:14:16 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240419-161123-6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae-pr5062.zip) |
| :heavy_check_mark: | 6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5062/4/) | 2024-04-19 16:15:16 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240419-161123-6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae-pr5062.zip) |
| :heavy_multiplication_x: | 6f6ab1c18b04ff42d2a373f7956d0b2d1e6dc6ae | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/4/) | 2024-04-19 16:48:17 | ~36 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 804c8f8ea9c1929a9b7484690bff16d6fefd58ce | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5062/5/) | 2024-04-19 16:58:30 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240419-165549-804c8f8ea9c1929a9b7484690bff16d6fefd58ce-pr5062.zip) |
| :heavy_check_mark: | 804c8f8ea9c1929a9b7484690bff16d6fefd58ce | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5062/5/) | 2024-04-19 16:59:10 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240419-165549-804c8f8ea9c1929a9b7484690bff16d6fefd58ce-pr5062.zip) |
| :heavy_check_mark: | 804c8f8ea9c1929a9b7484690bff16d6fefd58ce | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5062/5/) | 2024-04-19 17:02:18 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240419-165549-804c8f8ea9c1929a9b7484690bff16d6fefd58ce-pr5062.aar) |
| :heavy_multiplication_x: | 804c8f8ea9c1929a9b7484690bff16d6fefd58ce | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/5/) | 2024-04-19 17:35:35 | ~39 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/5//consoleText) |
| :heavy_multiplication_x: | 804c8f8ea9c1929a9b7484690bff16d6fefd58ce | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/6/) | 2024-04-19 18:42:53 | ~39 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/6//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b7e045e67b20643f142ac5217c8dc78143060b93 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5062/6/) | 2024-05-15 18:23:53 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240515-181938-b7e045e67b20643f142ac5217c8dc78143060b93-pr5062.zip) |
| :heavy_check_mark: | b7e045e67b20643f142ac5217c8dc78143060b93 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5062/6/) | 2024-05-15 18:24:19 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240515-181935-b7e045e67b20643f142ac5217c8dc78143060b93-pr5062.zip) |
| :heavy_check_mark: | b7e045e67b20643f142ac5217c8dc78143060b93 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5062/6/) | 2024-05-15 18:24:50 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240515-181935-b7e045e67b20643f142ac5217c8dc78143060b93-pr5062.aar) |
| | | | | | | |
| :heavy_check_mark: | b7e045e6 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/7/) | 2024-05-15 19:04:04 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5062/7//consoleText) |

osmaczko reviewed (commented):
Thanks for precise description.

As far as I can see, current fix proposal will not work at runtime. Meaning, if the app is already opened and the permission to channel is lost, unview counts will stay as they are.

> Also, keeping the DB value means that if the channel goes back to visible to us, we'll have a badge for the previous unread messages we had.

I am not sure about that assumption. The cache you modified will be eventually saved to the db and the values will be zeroed there as well.

> Another solution would be to change the DB value of the chat when we receive a new permission and detect that we no longer have view access. It's cleaner, but it doesn't fix previous account having the issue.

I think the issue is more complicated than assumed. `unviewed_message_count` and `unviewed_mentions_count` are denormalized fields and can be modified in [BlockContact](https://github.com/status-im/status-go/blob/d246699c5e1c19b3f905a6c10c3526d2004de273/protocol/message_persistence.go#L2133-L2136) for instance. So if you block contact that wrote a message on channel you have no permissions to, the unread counts will be back there.

I am thinking about the following solution:
1) Add `canView` column to chats' db
2) Modify the logic of denormalized fields recalculation, so that it zeroes when `!chat.canView` - this will ensure no surprises as described above
3) Whenever `chat.canView` is changed, trigger denormalized fields recalculation
4) Keep changes in `Init` as you did + update `chat.canView`

Ideally point 4) should be done through migration but since we keep community description marshaled I don't think it is even possible.

The points 1-3 are good candidates to be addressed in another PR I suppose.

jrainville commented:
> As far as I can see, current fix proposal will not work at runtime. Meaning, if the app is already opened and the permission to channel is lost, unview counts will stay as they are.

You're right.

> > Also, keeping the DB value means that if the channel goes back to visible to us, we'll have a badge for the previous unread messages we had.
>
> I am not sure about that assumption. The cache you modified will be eventually saved to the db and the values will be zeroed there as well.

Good point, I hadn't thought about it. 

> Ideally point 4) should be done through migration but since we keep community description marshaled I don't think it is even possible.

Indeed, I was thinking to see if there was a possible migration that would fix this in one pass and not think about it anymore, but it's impossible with how the Community description is marshalled.

> The points 1-3 are good candidates to be addressed in another PR I suppose.

I think to correctly fix the issue, I'll need to implement some of the points from 1-3 right away. Otherwise, like you said, we'll be missing the update on runtime.

Thanks for the great insight!

jrainville committed:
fix(messenger)_: make sure chats have an unread count of 0 for channels you can't view

Fixes https://github.com/status-im/status-desktop/issues/14421

The problem is that you can receive messages to  a channel, then later, before marking them as read, a permission is added to them, so you no longer have access.
Then, you can't even mark it as read if it's hidden.
Here, I fix it by setting the unread count on Init at 0 if the user doesn't have view access to it. And I make sure we update the counts when we are removed from a channel

jrainville commented:
Alright, I pushed a new commit that addresses the problem that when the channel becomes unavailable, the count wouldn't update.

I fixed it by checking if we are removed from a community channel, then we reset the counts to 0.

I'll keep the changes in Init too to fix people who had the problem happen to them before this fix (since we can't easily do a migration).

These changes don't address putting back the old count when the channel is back available, but IMO it's an edge case plus not that useful, so it's not worth recalculating.

As for the other points to improve, I think they will be mostly addressed in this refactor to align CommunityChats and Chats (adding canView, canPost, etc to the Chat) https://github.com/status-im/status-desktop/issues/14219

jrainville commented:
Now that you're back @osmaczko you can review again :smile:

osmaczko commented on protocol/messenger_communities.go:
```diff
@@ -3089,6 +3089,24 @@ func (m *Messenger) handleCommunityResponse(state *ReceivedMessageState, communi
+	// Check if we have been removed from a chat (ie no longer have access)
+	for id, changes := range communityResponse.Changes.ChatsModified {
+		if _, ok := changes.MembersRemoved[common.PubkeyToHex(&m.identity.PublicKey)]; ok {
+			chatID := community.IDString() + id
```
```suggestion
			chatID := community.ChatID(id)
```

osmaczko commented on protocol/messenger_communities.go:
```diff
@@ -3089,6 +3089,24 @@ func (m *Messenger) handleCommunityResponse(state *ReceivedMessageState, communi
 	}
 
+	// Check if we have been removed from a chat (ie no longer have access)
+	for id, changes := range communityResponse.Changes.ChatsModified {
```
```suggestion
	for channelID, changes := range communityResponse.Changes.ChatsModified {
```
